dynamic_programming/longest_common_subsequence.py

```diff
@@ -69,14 +72,73 @@ def longest_common_subsequence(x: str, y: str):
     return L[m][n], seq
 
 
+"""
+This is the Top-Down Approach (Memoization) of Dynamic Programming.
+"""
+
+
+def longest_common_subsequence_memoization(X, Y, m, n, dp):
+    """
+    :param X: String 1st
+    :param Y: String 2nc
+    :param m: length of String 1
+    :param n: length of String 2
+    :param dp: Array for Storage of The Pre Calculate Value
+    :return: length of Common Subsequence
+    """
+    """
+    >>> longest_common_subsequence_memoization("programming", "gaming")
+    6
+    >>> longest_common_subsequence_memoization("physics", "smartphone")
+    2
+    >>> longest_common_subsequence_memoization("computer", "food")
+    1
+    """
+    # base case
+    if m == 0 or n == 0:
+        return 0
+
+    # if the same state has already been
+    # computed
+    if dp[m - 1][n - 1] != -1:
+        return dp[m - 1][n - 1]
+
+        # if equal, then we store the value of the
+    # function call
+    if X[m - 1] == Y[n - 1]:
+
+        # store it in arr to avoid further repetitive
+        # work in future function calls
+        dp[m - 1][n - 1] = 1 + longest_common_subsequence_memoization(
+            X, Y, m - 1, n - 1, dp
+        )
+        return dp[m - 1][n - 1]
+
+    else:
+
```

Comment:
Lose the blank lines on 109, 116, and 118 for readability.

Reply:
Done Sir